File: packages/shared-types/src/valuation.ts
/**
 * Контракты оценки (ТЗ §10.1, §10.2, §12). Зеркало apps/engine/app/schemas/valuation.py.
 */
import type { Currency, ExpressValuationResult, Operation, Segment } from './index'

export interface ValuationRequest {
  adminUnitId?: number
  lon?: number
  lat?: number
  segment: Segment
  operation: Operation
  area: number
  period?: string
  floor?: number
  totalFloors?: number
}

export interface Comparable {
  id: number
  area: number
  price: number
  pricePerSqm: number
  distanceM: number
  publishedAt?: string | null
  weight: number
}

export interface Adjustment {
  factor: string
  description: string
  coefficient: number
}

export interface DetailedValuationResult extends ExpressValuationResult {
  pricePerSqm: number
  adminUnitName?: string | null
  period?: string | null
  comparables: Comparable[]
  adjustments: Adjustment[]
  methodology: string
}

export type { Currency }

File: apps/engine/app/schemas/valuation.py
"""Схемы оценки (ТЗ §10.1, §10.2, §12). Зеркало packages/shared-types/valuation.ts."""

from __future__ import annotations

from pydantic import BaseModel, Field


class ValuationRequest(BaseModel):
    admin_unit_id: int | None = Field(None, alias="adminUnitId")
    lon: float | None = None
    lat: float | None = None
    segment: str = "apartment"
    operation: str = "sale"
    area: float = Field(..., gt=0)
    period: str | None = None
    floor: int | None = None
    total_floors: int | None = Field(None, alias="totalFloors")

    model_config = {"populate_by_name": True}


class Comparable(BaseModel):
    id: int
    area: float
    price: float
    price_per_sqm: float = Field(alias="pricePerSqm")
    distance_m: float = Field(alias="distanceM")
    published_at: str | None = Field(None, alias="publishedAt")
    weight: float

    model_config = {"populate_by_name": True}


class Adjustment(BaseModel):
    factor: str
    description: str
    coefficient: float


class ExpressResult(BaseModel):
    value: float
    currency: str = "UAH"
    confidence: float
    comparables_count: int = Field(alias="comparablesCount")

    model_config = {"populate_by_name": True}


class DetailedResult(ExpressResult):
    price_per_sqm: float = Field(alias="pricePerSqm")
    admin_unit_name: str | None = Field(None, alias="adminUnitName")
    period: str | None = None
    comparables: list[Comparable] = []
    adjustments: list[Adjustment] = []
    methodology: str = ""
